File: engine/schedule.py
"""
Scheduler: Leitner boxes over words, plus a pattern layer.

Two decisions the research pushed us to:

  Spacing     Cepeda et al. found the best gap is roughly 10-20% of how long you
              want the memory to last. Boxes of 1, 2, 4, 8, 16 days cover a term.
  Interleave  The 2025 classroom study found mixed rule types beat blocked ones at
              eight weeks. So a session never puts two words of one pattern together.

The pattern layer is the part most spelling apps skip. A word failing twice is a word
problem. A pattern failing twice is a rule problem, so we pull in sibling words that
use the same rule rather than redrilling the one word she got wrong.
"""

from datetime import date, timedelta

BOX_DAYS = {1: 1, 2: 2, 3: 4, 4: 8, 5: 16}
PATTERN_TRIGGER = 2   # failures of one pattern before siblings get pulled in


class Scheduler:
    def __init__(self, words, today=None):
        self.words = {w["word"]: w for w in words}
        self.today = today or date.today()
        self.state = {w: {"box": 1, "due": self.today, "seen": 0, "wrong": 0}
                      for w in self.words}
        self.patterns = {}
        for w in self.words.values():
            for p in w["patterns"]:
                self.patterns.setdefault(p, {"members": [], "seen": 0, "wrong": 0})
                self.patterns[p]["members"].append(w["word"])

    # ---------- recording ----------

    def record(self, word, correct, diagnosis_patterns=()):
        st = self.state[word]
        st["seen"] += 1
        for p in set(self.words[word]["patterns"]) | set(diagnosis_patterns):
            ps = self.patterns.setdefault(p, {"members": [], "seen": 0, "wrong": 0})
            ps["seen"] += 1
            if not correct:
                ps["wrong"] += 1
        if correct:
            st["box"] = min(5, st["box"] + 1)
        else:
            st["wrong"] += 1
            st["box"] = 1
        st["due"] = self.today + timedelta(days=BOX_DAYS[st["box"]])
        return st

    def advance_to(self, day):
        self.today = day

    # ---------- selection ----------

    def weak_patterns(self):
        return sorted([p for p, s in self.patterns.items() if s["wrong"] >= PATTERN_TRIGGER],
                      key=lambda p: -self.patterns[p]["wrong"])

    def due(self):
        return [w for w, s in self.state.items() if s["due"] <= self.today]

    def session(self, size=10):
        """Reserve a third of the session for weak patterns, fill the rest with variety."""
        picked, reserve = [], max(1, size // 3)

        # 1. siblings of the patterns she keeps failing
        for p in self.weak_patterns():
            for sib in self.patterns[p]["members"]:
                if len(picked) < reserve and sib not in picked:
                    picked.append(sib)

        # 2. everything due, taken round-robin across patterns so no run of one rule
        due = sorted((w for w in self.due() if w not in picked),
                     key=lambda w: (self.state[w]["box"], -self.state[w]["wrong"], w))
        buckets = {}
        for w in due:
            buckets.setdefault(self.words[w]["patterns"][0], []).append(w)
        while len(picked) < size and any(buckets.values()):
            for key in list(buckets):
                if not buckets[key]:
                    del buckets[key]
                    continue
                picked.append(buckets[key].pop(0))
                if len(picked) >= size:
                    break
        return self._interleave(picked)

    def _interleave(self, words):
        """Greedy reorder so neighbouring words do not share a pattern."""
        remaining, out = list(words), []
        while remaining:
            last = set(self.words[out[-1]]["patterns"]) if out else set()
            nxt = next((w for w in remaining if not (set(self.words[w]["patterns"]) & last)),
                       remaining[0])
            remaining.remove(nxt)
            out.append(nxt)
        return out

    def report(self):
        boxes = {b: 0 for b in BOX_DAYS}
        for s in self.state.values():
            boxes[s["box"]] += 1
        return {"boxes": boxes, "weak_patterns": self.weak_patterns()[:5],
                "due_today": len(self.due())}

File: web/js/engine/schedule.js
// Port of engine/schedule.py. Same caveat as classify.js: mirror, do not improvise.
// tests/test_schedule_parity.py replays the Python golden file through this code.
//
// Leitner boxes of 1/2/4/8/16 days. docs/09 is honest that the intervals are a
// convention consistent with Cepeda et al. rather than derived from them: the optimal
// gap scales with the retention interval, roughly 10-20% of it, and there is no
// universal best gap.

export const BOX_DAYS = { 1: 1, 2: 2, 3: 4, 4: 8, 5: 16 };
export const PATTERN_TRIGGER = 2;

const DAY = 86400000;

export function addDays(iso, days) {
  return new Date(Date.parse(iso + "T00:00:00Z") + days * DAY).toISOString().slice(0, 10);
}

export function today() {
  return new Date().toISOString().slice(0, 10);
}

export class Scheduler {
  constructor(words, todayIso, state = null, patterns = null) {
    this.words = new Map(words.map((w) => [w.word, w]));
    this.today = todayIso || today();
    this.state = state || {};
    this.patterns = patterns || {};
    for (const w of this.words.values()) {
      if (!this.state[w.word]) {
        this.state[w.word] = { box: 1, due: this.today, seen: 0, wrong: 0 };
      }
      for (const p of w.patterns) {
        if (!this.patterns[p]) this.patterns[p] = { members: [], seen: 0, wrong: 0 };
        if (!this.patterns[p].members.includes(w.word)) this.patterns[p].members.push(w.word);
      }
    }
  }

  record(word, correct, diagnosisPatterns = []) {
    const st = this.state[word];
    st.seen += 1;
    const touched = new Set([...(this.words.get(word)?.patterns || []), ...diagnosisPatterns]);
    for (const p of touched) {
      if (!this.patterns[p]) this.patterns[p] = { members: [], seen: 0, wrong: 0 };
      this.patterns[p].seen += 1;
      if (!correct) this.patterns[p].wrong += 1;
    }
    if (correct) {
      st.box = Math.min(5, st.box + 1);
    } else {
      st.wrong += 1;
      st.box = 1;
    }
    st.due = addDays(this.today, BOX_DAYS[st.box]);
    return st;
  }

  advanceTo(iso) { this.today = iso; }

  weakPatterns() {
    // Python: sorted(list_of_keys, key=-wrong). Stable, so ties keep insertion order.
    return Object.keys(this.patterns)
      .filter((p) => this.patterns[p].wrong >= PATTERN_TRIGGER)
      .sort((a, b) => this.patterns[b].wrong - this.patterns[a].wrong);
  }

  due() {
    return Object.keys(this.state).filter((w) => this.state[w].due <= this.today);
  }

  session(size = 10) {
    const picked = [];
    const reserve = Math.max(1, Math.floor(size / 3));

    for (const p of this.weakPatterns()) {
      for (const sib of this.patterns[p].members) {
        if (picked.length < reserve && !picked.includes(sib)) picked.push(sib);
      }
    }

    const due = this.due()
      .filter((w) => !picked.includes(w))
      .sort((a, b) => {
        const sa = this.state[a], sb = this.state[b];
        return sa.box - sb.box || (-sa.wrong) - (-sb.wrong) || (a < b ? -1 : a > b ? 1 : 0);
      });

    const buckets = new Map();
    for (const w of due) {
      const key = this.words.get(w).patterns[0];
      if (!buckets.has(key)) buckets.set(key, []);
      buckets.get(key).push(w);
    }
    while (picked.length < size && [...buckets.values()].some((b) => b.length)) {
      for (const key of [...buckets.keys()]) {
        if (!buckets.get(key).length) { buckets.delete(key); continue; }
        picked.push(buckets.get(key).shift());
        if (picked.length >= size) break;
      }
    }
    return this.interleave(picked);
  }

  // Greedy reorder so neighbouring words do not share a pattern. docs/01 Tier 2:
  // interleaved beat blocked at eight weeks, effects 0.18 to 0.42, but with NO
  // transfer to untrained words. So interleave the practice and never assume the
  // rule generalised: that is what the off-list probe slice is for.
  interleave(words) {
    const remaining = [...words];
    const out = [];
    while (remaining.length) {
      const last = out.length
        ? new Set(this.words.get(out[out.length - 1]).patterns) : new Set();
      let idx = remaining.findIndex(
        (w) => !this.words.get(w).patterns.some((p) => last.has(p)));
      if (idx === -1) idx = 0;
      out.push(remaining.splice(idx, 1)[0]);
    }
    return out;
  }

  report() {
    const boxes = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0 };
    for (const s of Object.values(this.state)) boxes[s.box] += 1;
    return { boxes, weak_patterns: this.weakPatterns().slice(0, 5), due_today: this.due().length };
  }
}
